### ISSUE TITLE
LZ4_decompress_safe_partial_usingDict?

### ISSUE THREAD

**Issue by guillaumechereau** [lz4/lz4#225]
I would like to decompress a block using LZ4_decompress_safe_usingDict in several steps (like the LZ4_decompress_safe_partial would do).

Is there a way to do that using the current API?

**Comment by Cyan4973** [comment]
You can use `LZ4_decompress_safe_continue()`.

Setting the dictionary is done first with `LZ4_setStreamDecode()`

**Comment by guillaumechereau** [comment]
Thanks for the answer.

Actually I think I didn't formulated my question correctly.

What I am trying to do is split the decompression into several steps, so that I don't block the CPU doing it.  I work on an embedded system that receives big blocks of lz4 compressed data, each time I call a decompression function the main loop hangs on it.  One way to fix it would be to allow the decompression to be done in small steps.

So basically I am looking for a function similar to LZ4_decompress_safe_usingDict, that would return after a given number of sequences -within a single block- have been decoded, eventually decoding just a single sequence each time.

Does it make sense?

**Comment by Cyan4973** [comment]
OK !
Unfortunately, such function doesn't exist yet.

For such use case, my recommendation would be to cut data into smaller blocks.
That would be done from the compression side, so it requires to have some control over it.

**Comment by guillaumechereau** [comment]
Cutting in smaller blocks wouldn't work in my case either, since the streaming bandwidth is also limited so I need to compress the data as much as possible.
Anyway, thanks for the answers! I guess I will try to adapt your code to create my own decompression function.  If you have interest to add this to the API, I'll be happy to provide a pull request once I manage it.

**Comment by Cyan4973** [comment]
> Cutting in smaller blocks wouldn't work in my case either, since the streaming bandwidth is also limited so I need to compress the data as much as possible.

Well, when using `LZ4_compress_fast_continue()` or `LZ4_compress_HC_continue()`, blocks are not independent. Each block can still use previous 64 KB of data within previous block,
at least if blocks are laid next to each other.
So it should strongly reduce compression ratio losses associated with small blocks.

**Comment by guillaumechereau** [comment]
OK, that might be the solution for me then.  I will try this.  Thanks.

**Comment by Cyan4973** [comment]
has it solved your issue ?

**Comment by guillaumechereau** [comment]
I did some tests but unfortunately I couldn't get it to work properly,
so I gave up for the moment.  I'll definitively have to solve this at
some point, as right now my device has a small but noticeable delay
every time I uncompress a frame.

From an API point of view, I would still prefer to have a simple
stateless function where I systematically provide the pointer to the
dictionary.

**Comment by Cyan4973** [comment]
> From an API point of view, I would still prefer to have a simple stateless function where I systematically provide the pointer to the dictionary.

You mean, on the compression side ?
Because such API exists on the decompression side, it's `LZ4_decompress_safe_usingDict()`.

It would likely be possible to create something like `LZ4_compress_usingDict()` where dictionary would be manually specified. But compression will take a performance hit each time reloading the dictionary. Whereas this cost was basically zero with `LZ4_compress_fast_continue()`.

A big difference is that, with `LZ4_compress_fast_continue()`, the dictionary is "implicit", so it must remain at the same memory position it was seen before. Explicit allows to change that anytime.

**Comment by guillaumechereau** [comment]
On Sat, Sep 17, 2016 at 05:26:51AM -0700, Yann Collet wrote:
>You mean, on the compression side ?
>Because such API exists on the decompression side, it's
>`LZ4_decompress_safe_usingDict()`.

That is what I am using now, and it almost work.  What I am trying to do
is having exactly the same behavior than `LZ4_decompress_safe_usingDict`,
but in a way that could be split into several calls, each time
decompressing a limited number of sequences (if that makes sense).

By sequence I really mean one unit of the lz4 format: token, length,
literal, offset and match length.

Somehow I feel that it might be possible using the current API, but I
get a bit confused by the memory layout rules.

**Comment by Cyan4973** [comment]
You are right that this is possible (at least on the decompression side, compression side has some difficult subtleties to solve).
But unfortunately, not with current API.

Current API is hardwired on the concept of "block", so it wants to decompress a full block.
`LZ4_decompress_safe_partial()` is an exception, but it only allows an early-exit.
It doesn't support "resuming", which is the missing part to follow your idea.

It's technically possible to do it, but it will require writing a new function.

Still, I suspect that decoding smaller blocks will prove right to erase latency issues.
The issue is to learn how to create small inter-dependent blocks with current API.
I agree it's not trivial, there are issues to be aware of.

I recommend reading the excellent example from @t-mat :
[examples/blockStreaming_doubleBuffer.md](https://github.com/Cyan4973/lz4/blob/dev/examples/blockStreaming_doubleBuffer.md)
examples/blockStreaming_doubleBuffer.c

Consider typically a 2x64KB buffer methodology.
I believe it is a good example for your needs.
